domain/application/state/charm.go

```diff
@@ -40,37 +40,37 @@ func NewCharmState(factory database.TxnRunnerFactory) *CharmState {
 	}
 }
 
-// GetCharmIDByRevision returns the charm ID by the natural key, for a
-// specific revision.
+// GetCharmID returns the charm ID by the natural key, for a
+// specific revision and source.
 // If the charm does not exist, a [errors.CharmNotFound] error is returned.
-func (s *CharmState) GetCharmIDByRevision(ctx context.Context, name string, revision int) (corecharm.ID, error) {
+func (s *CharmState) GetCharmID(ctx context.Context, name string, revision int, source charm.CharmSource) (corecharm.ID, error) {
 	db, err := s.DB()
 	if err != nil {
 		return "", internalerrors.Capture(err)
 	}
 
-	var ident charmID
-	args := charmReferenceNameRevision{
+	var ident charmUUID
+	charmRef := charmReferenceNameRevisionSource{
 		ReferenceName: name,
 		Revision:      revision,
+		Source:        string(source),
 	}
 
 	query := `
-SELECT &charmID.*
-FROM charm
-INNER JOIN charm_origin
-ON charm.uuid = charm_origin.charm_uuid
-WHERE charm_origin.reference_name = $charmReferenceNameRevision.reference_name
-AND charm_origin.revision = $charmReferenceNameRevision.revision;
+SELECT &charmUUID.*
+FROM v_charm_origin
```

Comment:
Any other places this view could be used?

Reply:
It's already used on the app state to get the charm origin.